### PR TITLE
Implement encoder/decoder-based `truffle call` command

### PR DESCRIPTION
**Pull request by gnidan**
This draft PR adds a `truffle call` command leveraging Truffle's encoding and decoding functionality (including overload resolution :)
Screen Shot 2022-07-19 at 9 12 26 PM

### CODE REVIEW

File: packages/core/lib/commands/call/run.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  const contractNames = fs
+    .readdirSync(config.contracts_build_directory)
+    .filter(filename => filename.endsWith(".json"))
+    .map(filename => filename.slice(0, -".json".length));
+  const contracts = contractNames
+    .map(contractName => ({
+      [contractName]: config.resolver.require(contractName)
+    }))
+    .reduce((a, b) => ({ ...a, ...b }), {});
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:24]: I really hope this isn't the best way to list all artifacts...

File: packages/core/lib/commands/call/run.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  const settings = {
+    provider: config.provider,
+    projectInfo: {
+      artifacts: Object.values(contracts)
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:29]: Technically these are getting passed as contract abstraction objects, not as plain JSON artifacts, but it's probably fine.

File: packages/core/lib/commands/call/run.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    .readdirSync(config.contracts_build_directory)
+    .filter(filename => filename.endsWith(".json"))
+    .map(filename => filename.slice(0, -".json".length));
+  const contracts = contractNames
+    .map(contractName => ({
+      [contractName]: config.resolver.require(contractName)
+    }))
+    .reduce((a, b) => ({ ...a, ...b }), {});
+
+  const settings = {
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:26]: Will these settings differ for @truffle/encoder and @truffle/decoder? cc @haltman-at

File: packages/core/lib/commands/call/run.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    provider: config.provider,
+    projectInfo: {
+      artifacts: Object.values(contracts)
+    }
+  };
+
+  const contract = contracts[contractName];
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:33]: This needs handling for failure conditions - e.g. we might want to remind the user to run `truffle compile` first

File: packages/core/lib/commands/call/run.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  const instance = await contract.deployed();
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:34]: Similarly, this needs error-handling in case the user didn't run `truffle migrate` or what-have-you.

File: packages/core/lib/commands/call/meta.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+module.exports = {
+  command: "call",
+  description: "Call read-only contract function with arguments",
+  builder: {},
+  help: {
+    usage:
+      "truffle call <contract-name> <function-name> [<function-args...>] [--network=<network-name>",
```
Comment (gnidan) [inline packages/core/lib/commands/call/meta.js:7]: Although this just says `<contract-name>`, an initial release of this should include support for `--fetch-external`, and thus you should be allowed to specify either a contract name or an address.

File: packages/core/lib/commands/call/run.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+  const encoder = await Encoder.forContractInstance(instance, settings);
+
+  const { abi: functionEntry, tx: transaction } =
+    await encoder.encodeTransaction(functionName, args);
+
+  // wrap provider for lazy EIP-1193 compatibility
+  const provider = new Encoder.ProviderAdapter(config.provider);
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:42]: I'd advocate against merging until we figure out what to do about this.

For this quick hacking, I just wanted to do the EIP-1193 standard `provider.request`, so I sorta just... made the ProviderAdapter into an EIP-1193 provider (ish)

If we decide that this behavior is nice, then it probably should be a separate PR first to sort that out. 

(Of course, we could otherwise just use whatever gross syntax `config.provider.send()` gives us :)

File: packages/core/lib/commands/call/run.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+  const result = await provider.request({
+    method: "eth_call",
+    params: [transaction]
+  });
+
+  const decoder = await Decoder.forContractInstance(instance, settings);
+
+  const [decoding] = await decoder.decodeReturnValue(functionEntry, result);
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:51]: This will need to handle the cases where there is not exactly 1 decoding returned... obviously error nicely if there's 0, and then maybe give a warning if there's >1?

File: packages/core/lib/commands/call/run.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+  config.logger.log(
+    util.inspect(new Codec.Export.ReturndataDecodingInspector(decoding), {
+      colors: true,
+      depth: null,
+      maxArrayLength: null,
+      breakLength: 79
+    })
+  );
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:60]: We probably don't want to use `ReturndataDecodingInspector` here, since the output should be hand-crafted based on the context of the running command...

So instead of just outputting "Returned values: ...", we should probably use `ResultInspector` for each of the returned values and control how it looks tightly.

Comment (gnidan) [reply packages/core/lib/commands/call/run.js:60]: Note that there's some work here... we want to make sure we can handle situations where there's exactly one return value, vs. more than one, as well as handle situations where return values have names vs. not.

File: packages/core/lib/global-command-options.js [hunk 10]
```diff
@@ -4,6 +4,11 @@ const options = {
     description:
       "Specify the network to use. Network name must exist in the configuration."
   },
+  from: {
+    option: "--from <account>",
+    description:
+      "Specify the account from which to make calls or send transactions."
+  },
```
Comment (gnidan) [inline packages/core/lib/global-command-options.js:11]: Does this belong here? Should I just move this to a quick separate PR?

File: packages/core/lib/commands/call/run.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    }
+  };
+
+  const contract = contracts[contractName];
+  const instance = await contract.deployed();
+
+  const encoder = await Encoder.forContractInstance(instance, settings);
+
+  const { abi: functionEntry, tx: transaction } =
+    await encoder.encodeTransaction(functionName, args);
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:39]: Once the encoder makes its overload determination, this should probably check the `stateMutability` and throw a warning if it's not a readonly function

**Review by gnidan — COMMENTED** [review]
This was definitely just a quick hack, and I'm probably missing things that will need further consideration, but here's what I knew I was neglecting as I was writing this :)

File: packages/core/lib/commands/call/run.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+  // wrap provider for lazy EIP-1193 compatibility
+  const provider = new Encoder.ProviderAdapter(config.provider);
+
+  const result = await provider.request({
+    method: "eth_call",
+    params: [transaction]
```
Comment (gnidan) [inline packages/core/lib/commands/call/run.js:46]: Oh, oops, I forgot to hook up `--from` here!

Note that this currently just relies on @truffle/encoder producing this parameter comprehensively, and further relies on Ganache for filling in gaps with whatever defaults.

This should probably handle the various options more explicitly, or at least check to make sure this works against non-Ganache clients

**Comment by gnidan** [comment]
Oh, probably should also mention... once there are actual well-understood edge cases to test, this PR should probably also include tests for those :)

File: packages/core/lib/commands/call/meta.js (hunk 6 above)
Comment (gnidan) [reply packages/core/lib/commands/call/meta.js:7]: and for `<function-name>`, this should also allow the raw signature (e.g. `truffle call MyContract 'myFunction(string,uint)' ...`

the reason for this is: consider a contract that defines function `f` with two overloads (`f(string)` and `f(uint256)`). Today, if you pass a numeric string to the encoder, then the encoder's overload resolution system will always favor the latter overload, thus there's no good way currently to explicitly specify to use the string overload.

in addition, this situation probably even warrants a warning of some kind (e.g. "We notice you're calling a weirdly overloaded function... you might want to specify explicitly which overload you want")

File: packages/core/lib/commands/call/meta.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    usage:
+      "truffle call <contract-name> <function-name> [<args...>]\n" +
+      "                                " + // spacing to align with previous line
+      "[--network=<network-name>] [--from=<account>]",
+    options: [],
+    allowedGlobalOptions: ["from", "network", "config"]
```
Comment (gnidan) [inline packages/core/lib/commands/call/meta.js:13]: oh I just remembered... I also forgot about a blockNumber parameter (default `"latest"`)

Comment (gnidan) [reply packages/core/lib/commands/call/meta.js:13]: @haltman-at says "oh so you'll need CLI flags for every damn tx option"

Comment (gnidan) [reply packages/core/lib/commands/call/meta.js:13]: we can probably skip access lists tho

File: packages/core/lib/global-command-options.js (hunk 10 above)
Comment (sukanyaparashar) [reply packages/core/lib/global-command-options.js:11]: Yeah! @gnidan. I think this can be a separate PR since it is adding a new property for global options, which of course, can later be used by other commands if needed. I will make another PR for adding the `from` property to the global options list.

**Comment by sukanyaparashar** [comment]
Closing this PR as a new one is opened https://github.com/trufflesuite/truffle/pull/5647.
